crates/ra_ide/src/completion/complete_trait_impl.rs

```diff
@@ -47,22 +47,39 @@ use crate::{
 };
 
 pub(crate) fn complete_trait_impl(acc: &mut Completions, ctx: &CompletionContext) {
-    let trigger = ctx.token.ancestors().find(|p| match p.kind() {
+    let mut tokens = ctx.token.ancestors();
+    let trigger = tokens.find(|p| match p.kind() {
         SyntaxKind::FN_DEF
         | SyntaxKind::TYPE_ALIAS_DEF
         | SyntaxKind::CONST_DEF
+        | SyntaxKind::NAME_REF
         | SyntaxKind::BLOCK_EXPR => true,
         _ => false,
     });
 
-    let impl_def = trigger
-        .as_ref()
-        .and_then(|node| node.parent())
-        .and_then(|node| node.parent())
-        .and_then(ast::ImplDef::cast);
```

Comment:
Hm, could you check that this does not give completions if trigger happens to be inside function body? It seeems like this will complete not only top-level things, but anything nested into an impl.

Reply:
You were right here. I added a check to ensure there's no nesting.